### PR TITLE
tools: Add a Rust-based standard OCI container runtime based on Kata agent

### PR DESCRIPTION
**Pull request by ManaSugi**
Add a Rust-based standard OCI container runtime based on Kata agent.

Fixes: #2784

**Below is the README.**

### Change log

**Apr/28th/2022:**  Update to the latest README
**Nov/5th/2021:**  Mark seccomp support in the table of features compared to `runc`
**Oct/23th/2021:** Add `spec` command to the current status table
**Oct/13th/2021:** Add current status of `runk` compared to `runc`

# runk

## Overview

> **Warnings:**
> `runk` is currently an experimental tool.
> Only continue if you are using a non-critical system.

`runk` is a standard OCI container runtime written in Rust based on a modified version of
the [Kata Container agent](https://github.com/kata-containers/kata-containers/tree/main/src/agent), `kata-agent`.

`runk` conforms to the [OCI Container Runtime specifications](https://github.com/opencontainers/runtime-spec).

Unlike the [Kata Container runtime](https://github.com/kata-containers/kata-containers/tree/main/src/agent#features),
`kata-runtime`, `runk` spawns and runs containers on the host machine directly.
The user can run `runk` in the same way as the existing container runtimes such as `runc`,
the most used implementation of the OCI runtime specs.

## Why does `runk` exist?

The `kata-agent` is a process running inside a virtual machine (VM) as a supervisor for managing containers
and processes running within those containers.
In other words, the `kata-agent` is a kind of "low-level" container runtime inside VM because the agent
spawns and runs containers according to the OCI runtime specs.
However, the `kata-agent` does not have the OCI Command-Line Interface (CLI) that is defined in the
[runtime spec](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md).
The `kata-runtime` provides the CLI part of the Kata Containers runtime component,
but the `kata-runtime` is a container runtime for creating hardware-virtualized containers running on the host.

`runk` is a Rust-based standard OCI container runtime that manages normal containers,
not hardware-virtualized containers.
`runk` aims to become one of the alternatives to existing OCI compliant container runtimes.
The `kata-agent` has most of the [features](https://github.com/kata-containers/kata-containers/tree/main/src/agent#features)
needed for the container runtime and delivers high performance with a low memory footprint owing to the
implementation by Rust language.
Therefore, `runk` leverages the mechanism of the `kata-agent` to avoid reinventing the wheel.

## Performance

`runk` is faster than `runc` and has a lower memory footprint.

This table shows the average of the elapsed time and the memory footprint (maximum resident set size)
for running sequentially 100 containers, the containers run `/bin/true` using `run` command with
[detached mode](https://github.com/opencontainers/runc/blob/master/docs/terminals.md#detached)
on 12 CPU cores (`3.8 GHz AMD Ryzen 9 3900X`) and 32 GiB of RAM.
`runk` always runs containers with detached mode currently.

Evaluation Results:

|                       | `runk` (v0.0.1) | `runc` (v1.0.3) | `crun` (v1.4.2) |
|-----------------------|---------------|---------------|---------------|
| time [ms]           | 39.83         | 50.39         | 38.41         |
| memory footprint [MB] | 4.013         | 10.78         | 1.738         |

## Status of `runk`

We drafted the initial code here, and any contributions to `runk` and [`kata-agent`](https://github.com/kata-containers/kata-containers/tree/main/src/agent)
are welcome.

Regarding features compared to `runc`, see the `Status of runk` section in the [issue](https://github.com/kata-containers/kata-containers/issues/2784).

## Building

`runk` uses the modified the `kata-agent` binary, `oci-kata-agent`, which is an agent to be called from `runk`.
Therefore, you also need to build the `oci-kata-agent` to run `runk`.

You can build both `runk` and `oci-kata-agent` as follows.

```bash
$ cd runk
$ make
```

To install `runk` and `oci-kata-agent` into default directory for install executable program (`/usr/local/bin`):

```bash
$ sudo make install
```

## Using `runk` directly

Please note that `runk` is a low level tool not developed with an end user in mind.
It is mostly employed by other higher-level container software like `containerd`.

If you still want to use `runk` directly, here's how.

### Prerequisites

It is necessary to create an OCI bundle to use the tool. The simplest method is:

``` bash
$ bundle_dir="bundle"
$ rootfs_dir="$bundle_dir/rootfs"
$ image="busybox"
$ mkdir -p "$rootfs_dir" && (cd "$bundle_dir" && runk spec)
$ sudo docker export $(sudo docker create "$image") | tar -C "$rootfs_dir" -xvf -
```

> **Note:**
> If you use the unmodified `runk spec` template, this should give a `sh` session inside the container.
> However, if you use `runk` directly and run a container with the unmodified template,
> `runk` cannot launch the `sh` session because `runk` does not support terminal handling yet.
> You need to edit the process field in the `config.json` should look like this below
> with `"terminal": false` and `"args": ["sleep", "10"]`.

```json
"process": {
    "terminal": false,
    "user": {
        "uid": 0,
        "gid": 0
    },
    "args": [
        "sleep",
        "10"
    ],
    "env": [
        "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "TERM=xterm"
    ],
    "cwd": "/",
    [...]
}
```

If you want to launch the `sh` session inside the container, you need to run `runk` from `containerd`.

Please refer to the [Using `runk` from containerd](#using-runk-from-containerd) section

### Running a container

Now you can go through the [lifecycle operations](https://github.com/opencontainers/runtime-spec/blob/master/runtime.md)
in your shell.
You need to run `runk` as `root` because `runk` does not have the rootless feature which is the ability
to run containers without root privileges.

```bash
$ cd $bundle_dir

# Create a container
$ sudo runk create test

# View the container is created and in the "created" state
$ sudo runk state test

# Start the process inside the container
$ sudo runk start test

# After 10 seconds view that the container has exited and is now in the "stopped" state
$ sudo runk state test

# Now delete the container
$ sudo runk delete test
```

## Using `runk` from `containerd`

`runk` can run containers with the containerd runtime handler support on `containerd`.

### Prerequisites for `runk` with containerd

* `containerd` v1.2.4 or above
* `cri-tool`

> **Note:**
> [`cri-tools`](https://github.com/kubernetes-sigs/cri-tools) is a set of tools for CRI
> used for development and testing.

Install `cri-tools` from source code:

```bash
$ go get github.com/kubernetes-incubator/cri-tools
$ pushd $GOPATH/src/github.com/kubernetes-incubator/cri-tools
$ make
$ sudo -E make install
$ popd
```

Write the `crictl` configuration file:

``` bash
$ cat <<EOF | sudo tee /etc/crictl.yaml
runtime-endpoint: unix:///run/containerd/containerd.sock
EOF
```

### Configure `containerd` to use `runk`

Update `/etc/containerd/config.toml`:

```bash
$ cat <<EOF | sudo tee /etc/containerd/config.toml
version = 2
[plugins."io.containerd.runtime.v1.linux"]
  shim_debug = true
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
  runtime_type = "io.containerd.runc.v2"
[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runk]
  runtime_type = "io.containerd.runc.v2"
EOF
```

Restart `containerd`:

```bash
$ sudo systemctl restart containerd
```

### Running a container with `crictl` command line

You can run containers in `runk` via containerd's CRI.

Pull the `busybox` image:

``` bash
$ sudo crictl pull busybox
```

Create the sandbox configuration:

``` bash
$ cat <<EOF | tee sandbox.json
{
    "metadata": {
        "name": "busybox-sandbox",
        "namespace": "default",
        "attempt": 1,
        "uid": "hdishd83djaidwnduwk28bcsb"
    },
    "log_directory": "/tmp",
    "linux": {
    }
}
EOF
```

Create the container configuration:

``` bash
$ cat <<EOF | tee container.json
{
    "metadata": {
        "name": "busybox"
    },
    "image": {
        "image": "docker.io/busybox"
    },
    "command": [
        "sh"
    ],
    "envs": [
        {
            "key": "PATH",
            "value": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
        },
        {
            "key": "TERM",
            "value": "xterm"
        }
    ],
    "log_path": "busybox.0.log",
    "stdin": true,
    "stdin_once": true,
    "tty": true
}
EOF
```

With the `crictl` command line of `cri-tools`, you can specify runtime class with `-r` or `--runtime` flag.

Launch a sandbox and container using the `crictl`:

```bash
# Run a container inside a sandbox
$ sudo crictl run -r runk container.json sandbox.json
f492eee753887ba3dfbba9022028975380739aba1269df431d097b73b23c3871

# Attach to the running container
$ sudo crictl attach --stdin --tty f492eee753887ba3dfbba9022028975380739aba1269df431d097b73b23c3871
/ #
```

### CODE REVIEW

**Comment by ManaSugi** [comment]
I would like to discuss if the `runk` is of value to Kata community. 
Thank you.

**Comment by sameo** [comment]
Hey @ManaSugi, thanks for the PR. This is very exciting and it's nice to see `kata-agent` being used that way.

I guess my first question would be: How would you compare `runk` against [`youki`](https://github.com/containers/youki) which is arguably the leading rust based OCI runtime?

**Comment by devimc** [comment]
@ManaSugi thanks for raising this, would you like to  discuss this in the following Kata Architecture Commitee meeting ? 
https://etherpad.opendev.org/p/Kata_Containers_2021_Architecture_Committee_Mtgs

**Comment by ManaSugi** [comment]
Hello @sameo 
Thank you for your comment and question.
I think that the features for spawning and running containers that `kata-agent` and `youki` have are not so different.
`youki` OCI features: https://github.com/containers/youki#status-of-youki
`kata-agent` OCI features: https://github.com/kata-containers/kata-containers/tree/main/src/agent#features
(`kata-agent` is ready to support Seccomp feature #1788.)

So, if `runk` supports more CLI command operations such as `exec`, `ps`, etc. and we can prepare more tests, `runk` has a good chance of becoming a leading rust-based OCI runtime. Even if `runk` could not become the leading runtime, I guess that the number of users who use `kata-agent` through `runk` will increase. As a result, `runk` will lead to boost the development speed of `kata-agent`.

The only thing is that `runk` need to fork/exec `kata-agent` and it causes an additional overhead, whereas `youki` is one binary.
We can avoid the overhead by calling APIs inside `rustjail` directly. However, it is hard to achieve this without significantly changing the existing `kata-agent` code since `rustjail` has been developed on the assumption that it is called from `kata-agent`.

Hello @devimc 
Thank you for inviting me to the meeting. I would like to discuss `runk` there. Could you tell me if there is something I have to do to attend the meeting?

**Comment by devimc** [comment]
@ManaSugi add a new topic to the agent https://etherpad.opendev.org/p/Kata_Containers_2021_Architecture_Committee_Mtgs and join the meeting
Dial-In link: https://zoom.us/j/91444082924
PASSCODE: "kata"

**Comment by ManaSugi** [comment]
@devimc I added a topic to the agenda.

**Comment by dgibson** [comment]
Honestly, implementing an OCI runtime in terms of kata-agent, rather than just rustjail seems like an odd choice to me.  Most of what the kata agent (other than rustjail) deals with is essentially kata specific - how to translate devices, networks and mounts from outside the VM into the VM.  Likewise a lot of the information sent over the grpc link to the agent also only makes sense in the context of a VM.

File: tools/runk/README.md
Comment (c3d) [inline tools/runk/README.md]: @ManaSugi Could you clarify if the numbers in the table below are for `runk` by itself, or for `runk` + the `oci-kata-agent`?

Comment (ManaSugi) [reply tools/runk/README.md]: @c3d Thank you for the comment. The table shows `runk` + `oci-kata-agent` because `runk` calls `oci-kata-agent` internally for spawning and running a container.

**Comment by ManaSugi** [comment]
Resolved conflicts and updated `agent/Makefile` and `runk/Cargo.lock`

**Comment by ManaSugi** [comment]
I would appreciate it if anyone could review `runk`. If there is no problem, I'll squash the commits.

File: tools/runk/src/main.rs
Comment (jodh-intel) [inline tools/runk/src/main.rs:46]: Nit: "pseudo-terminal".

File: tools/runk/src/main.rs
Comment (jodh-intel) [inline tools/runk/src/main.rs:54]: Can you use clap's `crate_name!()`, `crate_version!()` and `crate_authors!()` macros here instead of hard-coding the values?

File: tools/runk/src/main.rs
Comment (jodh-intel) [reply tools/runk/src/main.rs:168]: Yes, but the problem is you have two pieces of code which guarantee state, but they are disconnected (and thus likely to deviate over time). We're trying to remove as many `unwrap()` and `expect()` calls as practical and this can be avoided with a slight reworking of the code.

File: tools/runk/src/spec.rs
Comment (jodh-intel) [inline tools/runk/src/spec.rs:26]: Might be worth making this a `const`.

File: tools/runk/src/start.rs
Comment (jodh-intel) [inline tools/runk/src/start.rs:22]: Why not log the current statu to make the error clearer?

File: tools/runk/src/state.rs
Comment (jodh-intel) [inline tools/runk/src/state.rs:69]: What if `status.id == ""`?

**Review by jodh-intel — COMMENTED** [review]
Thanks @ManaSugi. I've done a quick review but will look at this more closely on Monday.

**Comment by ManaSugi** [comment]
@jodh-intel I appreciate for the review and I'll fix them. 
In addition, I'll modify CLI component of `runk` using `clap` [Derive Macro](https://github.com/clap-rs/clap#using-derive-macros) for the future use cases as mentioned in [this comment](https://github.com/kata-containers/kata-containers/issues/2784#issuecomment-969200057).

**Comment by jodh-intel** [comment]
@ManaSugi - are you still planning to update this PR?

**Comment by ManaSugi** [comment]
@jodh-intel I apologize for the late update. I'll update this PR next week.

**Comment by jodh-intel** [comment]
Hi @ManaSugi - any update on this?

**Comment by jodh-intel** [comment]
PR marked as stale.

**Comment by ManaSugi** [comment]
@jodh-intel  Sorry for the late update, I'll update this using [`liboci-cli`](https://github.com/containers/youki/tree/main/crates/liboci-cli) as command line interface of `runk` for improving maintainability by the end of the week. In addtion, I'll squash the commits at the same time.

File: src/agent/rustjail/src/container.rs
Comment (ManaSugi) [reply src/agent/rustjail/src/container.rs]: Thank you for the review.
I removed all `unwrap()` method except for test.

File: tools/runk/Cargo.toml
Comment (ManaSugi) [reply tools/runk/Cargo.toml]: I fixied it.

File: tools/runk/Cargo.toml
Comment (ManaSugi) [reply tools/runk/Cargo.toml]: I fixed it.

File: tools/runk/README.md
Comment (ManaSugi) [reply tools/runk/README.md]: I fixed it.

File: tools/runk/README.md
Comment (ManaSugi) [reply tools/runk/README.md]: I fixed it.

File: tools/runk/src/cgroup.rs
Comment (ManaSugi) [reply tools/runk/src/cgroup.rs]: Thank you for the review.
I removed my`remove_dir` function.
Regarding `sleep`, I use it for a conservative approach to delete cgroups path. If container processes exist in the cgroups, `runk` cannot delete the cgroups directory due to a `Device or resource busy` error. Before calling `remove_cgroup_dir`, `runk` kills the container process, but it may take some time for the process to finish. Hence, `runk` uses `sleep` approach to wait for the process to finish without an infinite loop.
`runc` also takes the same approach:
https://github.com/opencontainers/runc/blob/6ff042023ce8f550a70d3c0bd2588e48df70d939/libcontainer/cgroups/utils.go#L270

File: tools/runk/src/container.rs
Comment (ManaSugi) [reply tools/runk/src/container.rs]: I fixed it.

File: tools/runk/src/main.rs
Comment (ManaSugi) [reply tools/runk/src/main.rs]: I fixed it.

File: tools/runk/src/main.rs
Comment (ManaSugi) [reply tools/runk/src/main.rs]: I fixed it.

File: tools/runk/src/spec.rs
Comment (ManaSugi) [reply tools/runk/src/spec.rs]: I fixed it.

File: tools/runk/src/state.rs
Comment (ManaSugi) [reply tools/runk/src/state.rs]: When launching container using `runk`, it is mandatory to input the `container id` (empty string is prohibited) in the CLI, so I think that `status.id` won't be empty.

**Comment by ManaSugi** [comment]
@jodh-intel I apoplogize for the late update.
I redesigned `runk` using [`liboci-cli`](https://github.com/containers/youki/tree/main/crates/liboci-cli) as the cli of `runk` and cleaned the commits.

File: tools/runk/src/start.rs
Comment (ManaSugi) [reply tools/runk/src/start.rs]: I fixed it.

File: tools/runk/src/main.rs
Comment (ManaSugi) [reply tools/runk/src/main.rs]: I understand. I removed all `unwrap()` except for test.

File: src/tools/runk/Cargo.toml
Comment (utam0k) [inline src/tools/runk/Cargo.toml:14]: Hi! I am a maintainer of youki. I have the plan to publish `liboci-cli` to crates.io soon.
https://github.com/containers/youki/issues/586

Comment (ManaSugi) [reply src/tools/runk/Cargo.toml:14]: Thank you for the information. It's awesome!

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

**Comment by jodh-intel** [comment]
Any update on this @ManaSugi? Is it ready for re-review?

**Comment by ManaSugi** [comment]
@jodh-intel I have no updates. If you don't mind, could you re-review it?. Thank you.

File: src/tools/runk/Cargo.toml
Comment (jodh-intel) [reply src/tools/runk/Cargo.toml:14]: Any update on publishing on crates.io @utam0k?

File: src/tools/runk/README.md
Comment (jodh-intel) [inline src/tools/runk/README.md:19]: Nit: Overly long line.

File: src/tools/runk/README.md
Comment (jodh-intel) [inline src/tools/runk/README.md:44]: It might be worth adding version details, architectures and hardware tested on here.

File: src/tools/runk/README.md
Comment (jodh-intel) [inline src/tools/runk/README.md:49]: I would consider moving this section to the top of the file and making it a ["warning"](https://github.com/kata-containers/kata-containers/blob/main/docs/Documentation-Requirements.md#warnings-and-other-admonitions). Something like:

> **Warning:** `runk` is currently an experimental tool.

File: src/tools/runk/README.md
Comment (jodh-intel) [inline src/tools/runk/README.md:58]: I'd add that text into the warning at the top of the file.

File: src/tools/runk/README.md
Comment (jodh-intel) [inline src/tools/runk/README.md:135]: I think you should explain why it is necessary to run `runk` as `root`.

**Review by jodh-intel — COMMENTED** [review]
Thanks @ManaSugi - just a few doc comments...

**Comment by jodh-intel** [comment]
@ManaSugi - one other thing, if and when this lands, how is it going to be tested? Do you plan to create some integration tests in https://github.com/kata-containers/tests? If so, you might want to speak to @GabyCT and @dborquez as they can probably help. We removed a lot of the old Docker tests when we moved to the [shimv2 architecture](https://github.com/kata-containers/kata-containers/tree/main/docs/design/architecture#shim-v2-architecture) but these could be re-used for `runk` possibly.

**Comment by utam0k** [comment]
@jodh-intel 
I've been a little busy with my personal life and haven't had much time, but I have a plan. Can you please wait a little longer?
https://github.com/kata-containers/kata-containers/pull/2785#discussion_r807816694

**Comment by jodh-intel** [comment]
Hi @utam0k - Sure! I was just wondering :smile:

File: src/tools/runk/README.md
Comment (ManaSugi) [reply src/tools/runk/README.md]: I fixed it.

File: src/tools/runk/README.md
Comment (ManaSugi) [reply src/tools/runk/README.md]: I fixed it.

File: src/tools/runk/README.md
Comment (ManaSugi) [reply src/tools/runk/README.md]: I fixed it.

File: src/tools/runk/README.md
Comment (ManaSugi) [reply src/tools/runk/README.md]: I fixed it.

File: src/tools/runk/README.md
Comment (ManaSugi) [reply src/tools/runk/README.md:141]: I fixed it.

**Comment by ManaSugi** [comment]
@jodh-intel Thank you for your review. I fixed them.

> @ManaSugi - one other thing, if and when this lands, how is it going to be tested? Do you plan to create some integration tests in https://github.com/kata-containers/tests? If so, you might want to speak to @GabyCT and @dborquez as they can probably help. We removed a lot of the old Docker tests when we moved to the [shimv2 architecture](https://github.com/kata-containers/kata-containers/tree/main/docs/design/architecture#shim-v2-architecture) but these could be re-used for `runk` possibly.

Yes, I plan to create integration tests in https://github.com/kata-containers/tests for runk. Currently, runk cannot run with Docker, but when runk starts support for Docker, I'll try to re-use the old Docker tests. Thank you.

**Review by jodh-intel — APPROVED** [review]
Thansk @ManaSugi.

lgtm

**Comment by jodh-intel** [comment]
Ping for reviews - this PR has been open a very long time so would be good to land it folks...

File: src/tools/runk/README.md
Comment (jodh-intel) [inline src/tools/runk/README.md:160]: Hi @ManaSugi - this is causing the CI to fail since there are two identical heading names in the same document. You could change this to "### Prerequisites for runk with containerd" or similar?

**Comment by jodh-intel** [comment]
Can we get another review over here folks? This is a big PR and would be great to see this land. It already has one ack, so we need one more ack to land this new runtime!

@liubin, @bergwolf, @lifupan - could one of you tal?

File: src/tools/runk/README.md
Comment (ManaSugi) [reply src/tools/runk/README.md]: Thank you for the suggestion. I fixed it.

**Comment by utam0k** [comment]
Youki team has released the new version, and `liboci-cli` appeared on crates.io.
https://crates.io/crates/liboci-cli

**Comment by arronwy** [comment]
Can one of the admins verify this patch?

**Comment by ManaSugi** [comment]
I solved the conflicts and added 4 commits.

**Comment by zvonkok** [comment]
@ManaSugi Does `runk` support OCI runtime spec hooks?

**Comment by ManaSugi** [comment]
@zvonkok Yes, `runk` can use OCI hooks since `kata-agent` already supports the hooks, `prestart`, `poststart` and `poststop`.
(`prestart` is deprecated currently, so I have plan to add `createRuntime`, `createContainer`, `startContainer` to the agent instead.)

**Comment by jodh-intel** [comment]
Please can someone else tal at this PR?

**Review by fidencio — APPROVED** [review]
Let's have this one in for the 2.5.0-alpha1.

Thanks @ManaSugi!

**Comment by ManaSugi** [comment]
I appreciate for the review. If there is no problem, I'd like to squash the commits into initial two commits (e30ca624c7b60b85ccd99e96b7e4821615aa96e4 and e02c0ac60763393b24627f1531c981a9c1b17d2c).

**Comment by jodh-intel** [comment]
@ManaSugi - wfm, thanks :smile: I've added the dnm label to allow you to do that.

**Comment by ManaSugi** [comment]
@jodh-intel Thank you for adding the label. I squashed the commits.

File: src/agent/rustjail/src/container.rs
Comment (liubin) [inline src/agent/rustjail/src/container.rs:396]: Can this line be moved to before line 692?

File: src/tools/runk/README.md
Comment (liubin) [inline src/tools/runk/README.md:199]: `io.containerd.runc.v2` or `io.containerd.runk.v2` ?

File: src/tools/runk/src/main.rs
Comment (liubin) [inline src/tools/runk/src/main.rs:110]: Is this required?

**Comment by jodh-intel** [comment]
Thanks @ManaSugi !

File: src/tools/runk/libcontainer/src/cgroup.rs
Comment (liubin) [inline src/tools/runk/libcontainer/src/cgroup.rs:39]: Nit: paths -> path

**Comment by jodh-intel** [comment]
Hi @liubin - apologies, I hadn't seen your comments before I hit the big green button :smile:

@ManaSugi - Could you possibly raise a follow-up PR to address @liubin's comments?

**Pull request merged**
